u-boot: PCIe devices support for Raspberry PI5

oleksiimoisieiev opened the pull request:
Enhance support for the Raspberry PI 5 board.
This includes the following features:
1) Add separate configuration and device-tree files to support Raspberry PI 5 board. Previous implementation was based on the default rpi_arm64_defconfig
2) Add PCIe device support for BCM2712 board
3) Add RP1 driver to support the peripheral connected to the RP1 Chip (the new feature of the raspberrypi 5)
4) Add Clock, gpio, reset drivers for RP1 peripheral
5) Add support of MACB ethernet for BCM2712

**NOTE**: This PR includes a couple of "Hacks" that are not intended to be upstreamed and were implemented as a temprorary solution before linux RP1 driver gets upstreamed.

oleksiimoisieiev committed:
arch: arm: mach-bcm283x: add pcie memory region to BCM2712 memmap

Update initial memory map for the BCM2712 board. Add PCIe memory
regions.

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

oleksiimoisieiev committed:
drivers: clk: introduce clock driver for RP1

Introduce clock driver for RPI5 RP1 chip. It is intended to
work with a clock controller connected to the RP1 chip of RPI5.
Current implementation supports only RP1_CLK_ETH_TSU clock.
Additional clock could be added to the rp1_data array if needed.

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

oleksiimoisieiev committed:
drivers: gpio: add support of RP1 GPIO for Raspberry PI 5

Raspberry PI5 has an external chip RP1 installed which provides
it's own GPIO controller. This driver implements GPIO support
for RP1 GPIO controller.

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

oleksiimoisieiev committed:
drivers: pci: add BCM2712 support for pcie_brcmstb driver

Introducing pcie support for BCM2712 on example of the RPI5 board.
This driver initializes PCIe driver on start and provides access to
the peripheral, connected to PCIe.

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

oleksiimoisieiev committed:
drivers: net: macb: introduce ePCI connection support for macb

On BCM2712 connected devices, such as RPI5 has an external chip called
RP1 connected to the main SOC over ePCI. Macb is connectet to the RP1
chip and should be transparent to the SOC. Introduced changes are
intended to solve mapping issues and implement full support for the
macb driver.

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

oleksiimoisieiev committed:
drivers: reset: introduce reset drivers for brcmstb

Introducing Generic Reset controller and Generic RESCAL
drivers for Broadcom boards.

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

oleksiimoisieiev committed:
include: dt-bindings: clk: introduce bindings for rp1 clock

Introducing device-tree bindings for the RP1 clocks.

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

oleksiimoisieiev committed:
drivers: mfd: introduce RP1 chip driver for RPI5

Raspberry PI 5 has RP1 chip connected over PCIE and working as a
bridge to different hardware.
This driver implementation was inspired by MFD RP1 driver in the Linux
Kernel.

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

oleksiimoisieiev committed:
include: dt-bindings: mfd: introduce bindings for RP1 driver

Header includes all needed defines to be used in the device-tree
bindings.

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

oleksiimoisieiev committed:
arch: arm: mach-bcm283x: add BCM2712 board support

Add BCM2712 board support to U-Boot.

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

oleksiimoisieiev committed:
drivers: core: of_addr: fix of_get_dma_range translation

According to the Section 2.3.9 of the Device Tree specification [0]
the format of the value of the dma-ranges property is an arbitrary
number of triplets of (child-bus-address,
parent-bus-address, length).
Where:
* The child-bus-address is a physical address within the child bus’
  address space. The number of cells to represent the address depends on
  the bus and can be determined from the #address-cells of this node
  (the node in which the dma-ranges property appears).
* The parent-bus-address is a physical address within the parent bus’
  address space. The number of cells to represent the parent address
  is bus dependent and can be determined from the #address-cells
  property of the node that defines the parent’s address space.
* The length specifies the size of the range in the child’s address
  space. The number of cells to represent the size can be determined
  from the #size-cells of this node (the node in which the dma-ranges
  property appears)

This description matches comments in the of_addr source code. But
actual implementation do not match the specification.
'of_match_bus' call returns address-cells and size-cells of the parent
device. That's why the following calls were used to receive addr_cells
and size_cells of the current device:
- of_simple_addr_cells
- of_simple_size_cells

This fix was tested on RPI5 board with both device types: directly
connected (e.g Serial) and connected over PCIe (e.g ethernet).

For example when dma_ranges are parsing the following device tree
structure:
/ {
	pcie2: {
		#address-cells = <3>;
		#size-cells = <2>;
		// 64GB system RAM space at PCIe 10_00000000
		dma-ranges =
			   <0x02000000 0x00 0x00000000
			    0x1f 0x00000000
			    0x00 0x00400000>,
			   <0x43000000 0x10 0x00000000
			    0x00 0x00000000
			    0x10 0x00000000>;
		rp1: {
			#address-cells = <2>;
			#size-cells = <2>;
			dma-ranges =
			// inbound RP1 1x_xxxxxxxx -> PCIe 1x_xxxxxxxx
			  <0x10 0x00000000
			  0x43000000 0x10 0x00000000
			  0x10 0x00000000>,

			// inbound RP1 c0_40xxxxxx -> PCIe 00_00xxxxxx
			  <0xc0 0x40000000
			  0x02000000 0x0 0x00000000
			  0x0 0x00400000>,

			// inbound RP1 0x_xxxxxxxx -> PCIe 1x_xxxxxxxx
			  <0x00 0x00000000
			  0x02000000 0x10 0x00000000
			  0x10 0x00000000>;
		};
	};
};

So when we parse ranges on rp1 - #address-cells of the rp1 node should
be taken into account.
In the provided example we have the following format:
pcie2:
   #addres-cells = <0x3>;
   #size-cells = <0x2>;
   dma-ranges = < addr_hi addr_mid addr_lo
                  trans_addr_hi trans_addr_lo
		  size_hi size_lo>;

rp1:
   #addres-cells = <0x2>;
   #size-cells = <0x2>;
   dma-ranges = < addr_hi addr_lo
                  trans_addr_hi trans_addr_mid trans_addr_lo
		  size_hi size_lo>;

of_get_dma_range searches for the first dev with dma_ranges
and then executes of_match_bus which follows by count_cells call
for both device and parent.

count_cells callback returns #address-cells for the parent device to
the caller.
So in the provided example address cells should be returned 2 for rp1,
but we get 3 when using count_cells callback.
This breaks address translation and was fixed in the provided patch
by using of_simple_addr_cells call. Which will return correct
address-cells for the node. This matches the requirements of the
device-tree specification.

[0] https://readthedocs.org/projects/devicetree-specification/downloads/pdf/latest/

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc committed:
board: raspberrypi: rpi: request RP1 in late_init

RP1 chip is the PCIe device which should provide transparent access to
the peripheral connected to it for the main Soc.
Current RP1 device tree node structure and RP1 driver implementation
don't allow PCIe subsystem to make a proper initialization of RP1
device with childs. So RP1 probe should be called after all
initializations were done.

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc committed:
drivers: pci: take into account that ofnode_read_pci_vendev can fail

Prior to this fix devfn was set to 0, which lead to funny consequences
because PCI subsystem tried to use RP1 driver as a host bridge driver.

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

oleksiimoisieiev committed:
HACK: drivers: mfd: set bar configuration for RP1 driver

Set bar address in the correct order. RP1 driver in Linux Kernel rely on
the pci BAR configuration which initializes BARs based on BAR size.
This results the Linux set BARs in the same order on each boot.
U-boot does initialization without any sorting so the configuration
may be different from the Linux kernel.
Here BAR address configuration is set to match the Linux Kernel order
to avoid fails on address translation.
This should be changed after upstreaming RP1 driver to the Linux kernel
mainline.

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

GrygiriiS commented:
Fast look:P

1) rpi5 doesn't need DT in u-boot, so why do you need it? DT picked up from RPI5 bootloader by using mem address on FDT
2)  u-boot has "fixed-clock" driver support - mo need for get_macb_pclk_rate()
3) why rpi_arm64_defconfig is not ok for you?

And this is very questionable ;) "drivers: core: of_addr: fix of_get_dma_range translation"  at minimum commit message should contain full information describing what is not working with DT examples 

And you can't introduce "config BCM2712" and put it all over the code.

oleksiimoisieiev committed:
configs: add support for the Raspberrypi 5 board to default config

Add the default configuration file for RPI5 board.
To build u-boot for Raspberrypi 5 the following command should be
used:
> make rpi_arm64_defconfig CROSS_COMPILE=aarch64-linux-gnu-

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

oleksiimoisieiev commented:
> Fast look:P
> 
>     1. rpi5 doesn't need DT in u-boot, so why do you need it? DT picked up from RPI5 bootloader by using mem address on FDT

[OM]Fixed as discussed. 

> 
>     2. u-boot has "fixed-clock" driver support - mo need for get_macb_pclk_rate()

[OM]Thanks - i will take a look.

> 
>     3. why rpi_arm64_defconfig is not ok for you?
> 

[OM]Fixed as discussed. 

> 
> And this is very questionable ;) "drivers: core: of_addr: fix of_get_dma_range translation" at minimum commit message should contain full information describing what is not working with DT examples

[OM]Hm. It includes quote from the device-tree standard with comments why previous implementation didn't match. examples can be found on the provided link for the documentation. But I will update the desctiption.

> 
> And you can't introduce "config BCM2712" and put it all over the code.

[OM] That's how it was done in Linux. That's how it was done for AT91FAMILY in macb. The problem is that 2711 is completely different from 2712. The only solution a can think of is to use model structure filled by "get_board_revision"

lorc commented on arch/arm/mach-bcm283x/init.c:
```diff
@@ -165,6 +165,27 @@ static void rpi_update_mem_map(void)
+{
+	int len;
+	const int *cell;
+	int offset = fdt_path_offset(gd->fdt_blob, "/clocks/macb_pclk");
```
This puzzles me. Why you can't use standard clk API?

oleksiimoisieiev replied:
removed this code in terms of fixing comments from Grygorii. No need to implement arch clock when we have fixed-clock driver in system.

lorc commented on arch/arm/mach-bcm283x/init.c:
```diff
@@ -165,6 +165,27 @@ static void rpi_update_mem_map(void)
+	const int *cell;
+	int offset = fdt_path_offset(gd->fdt_blob, "/clocks/macb_pclk");
+
+	if (!offset)
```
It would be nice to print some error message

oleksiimoisieiev replied:
same as above. removed

lorc commented on arch/arm/mach-bcm283x/init.c:
```diff
@@ -165,6 +165,27 @@ static void rpi_update_mem_map(void)
+
+	cell = fdt_getprop(gd->fdt_blob, offset, "clock-frequency", &len);
+
+	if (!cell || len < sizeof(int))
```
It would be nice to print some error message

oleksiimoisieiev replied:
same as above. removed

lorc commented on drivers/mfd/Kconfig:
```diff
@@ -2,3 +2,13 @@ config MFD_ATMEL_SMC
+
+	  This driver provides support for the Raspberry Pi RP1 peripheral chip.
+	  It is responsible for enabling the Device Tree node once the PCIe endpoint
+	  has been configured, and handling interrupts.
```
Missing EOL at EOF

oleksiimoisieiev replied:
weird. checkpatch told me nothing about that. Fixed

lorc commented on drivers/mfd/rp1.c:
```diff
@@ -0,0 +1,133 @@
+	phys_addr_t bar_start;
+};
+
+static struct rp1_dev *g_rp1;
```
Do you really need the global pointer?

oleksiimoisieiev replied:
removed

lorc commented on drivers/mfd/rp1.c:
```diff
@@ -0,0 +1,133 @@
+};
+
+static struct rp1_dev *g_rp1;
+static u32 g_chip_id, g_platform;
```
Is this used anywhere?

oleksiimoisieiev replied:
removed

lorc commented on include/mfd/rp1_platform.h:
```diff
@@ -0,0 +1,23 @@
+#include <linux/types.h>
+#include <asm/byteorder.h>
+
+#define RP1_B0_CHIP_ID 0x10001927
```
I saw the same in rp1.c

oleksiimoisieiev replied:
removed rp1_platform.h. It was moved from linux source code, but I see no reason to leave it here

lorc commented on board/raspberrypi/rpi/rpi.c:
```diff
@@ -588,3 +588,19 @@ int ft_board_setup(void *blob, struct bd_info *bd)
 
 	return 0;
 }
+
```
I am not sure if this is a HACK. I saw exactly the same code for other platforms. I assume this is normal for U-BOOT

oleksiimoisieiev replied:
I agree that this may be proper solution for u-boot. I was thinking about rp1 driver itself. I suppose it should be moved out from this call once we get is included into the pcie device chain.

GrygiriiS replied:
Yep. It's probably not a HACK.

oleksiimoisieiev replied:
Ok. I will remove HACK tag but will leave a NOTE with a few words if you don't mind

oleksiimoisieiev replied:
removed

lorc commented on drivers/mfd/rp1.c:
```diff
@@ -74,6 +74,20 @@ static int rp1_probe(struct udevice *dev)
+	 * This should be changed after upstreaming RP1 driver to the Linux kernel
+	 * mainline.
+	 */
+	dm_pci_write_config32(dev, PCI_BASE_ADDRESS_1, 0);
```
I believe, in my setup all was working without this code. Could you please check?

oleksiimoisieiev replied:
unfortunatelly no. It doesn't work without this fix

lorc replied:
Okay then. We now have procedure for HACKs: https://xen-troops.github.io/process/process under "Temporary changes" section. Please follow it.

oleksiimoisieiev replied:
updated according to the process description

oleksiimoisieiev committed:
drivers: net: macb: do not include arch/clk.h when clocks enabled

During initialization MACB should request clock rates to work
properly. If CONFIG_CLK is not set then it uses special arch call
named: 'get_macb_pclk_rate'
This function is arch specific and should return fixed rate for macb
clock even if clock subsystem was disabled.
When CONFIG_CLK is enabled there is no need to include arch specific
code and no need to declare 'get_macb_pclk_rate' call.
Added conditional to include arch/clk.h only if clock subsystem is
disabled.

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

oleksiimoisieiev commented:
Fixed all comments. The only thing left is "And you can't introduce "config BCM2712" and put it all over the code." I'm still not sure how this could be done better. So any ideas will be appreciated.

oleksiimoisieiev committed:
board: raspberrypi: rpi: save board_type to the global_data

Save board_type retrieved from the BCM board firmware to the global
data. This information can be accessed by using gd_board_type() call to
determine the type of currently used board.

Signed-off-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

oleksiimoisieiev commented:
@lorc  @GrygiriiS  I have applied recommendation of Grygorii to avoid relying just on BCM2712 configuration and did the refactoring. Right now I think I'we managed to fix all comments to now it's ready to the second review stage.

GrygiriiS commented on configs/rpi_arm64_defconfig:
```diff
@@ -56,3 +57,28 @@ CONFIG_SYS_WHITE_ON_BLACK=y
+
+CONFIG_OF_LIVE=y
+
+CONFIG_DM=y
```
Not needed.

oleksiimoisieiev replied:
+

GrygiriiS commented on configs/rpi_arm64_defconfig:
```diff
@@ -56,3 +57,28 @@ CONFIG_SYS_WHITE_ON_BLACK=y
+CONFIG_DM=y
+
+CONFIG_MACB=y
+CONFIG_DM_ETH=y
```
Not needed

oleksiimoisieiev replied:
+

GrygiriiS commented on configs/rpi_arm64_defconfig:
```diff
@@ -56,3 +57,28 @@ CONFIG_SYS_WHITE_ON_BLACK=y
+CONFIG_MACB=y
+CONFIG_DM_ETH=y
+CONFIG_CMD_NET=y
+CONFIG_PHYLIB=y
```
Not needed.

oleksiimoisieiev replied:
+

GrygiriiS commented on configs/rpi_arm64_defconfig:
```diff
@@ -56,3 +57,28 @@ CONFIG_SYS_WHITE_ON_BLACK=y
+
+CONFIG_MACB=y
+CONFIG_DM_ETH=y
+CONFIG_CMD_NET=y
```
Not needed.

oleksiimoisieiev replied:
+

GrygiriiS commented on arch/arm/mach-bcm283x/init.c:
```diff
@@ -76,15 +76,34 @@ static struct mm_region bcm2712_mem_map[MEM_MAP_MAX_ENTRIES] = {
-	}, {
+	},
+	{
+		/* First 1GB of DRAM */
```
It's not first 1G and not PCIe. And it's not clear why is it needed?

oleksiimoisieiev replied:
removed

GrygiriiS commented on drivers/net/macb.c:
```diff
@@ -39,7 +39,9 @@
 #include <linux/mii.h>
 #include <asm/io.h>
 #include <linux/dma-mapping.h>
+#if !defined(CONFIG_CLK)
```
It's some leftovers I assume and should be dropped completely.

oleksiimoisieiev replied:
Oh, that's was tricky to find. Macb using arch_get_rate callback if clocks are disabled. but this ifdef was missing.

GrygiriiS replied:
ok

GrygiriiS commented on drivers/core/of_addr.c:
```diff
@@ -374,16 +374,16 @@ int of_get_dma_range(const struct device_node *dev, phys_addr_t *cpu,
-	bus_node = of_match_bus((struct device_node*)dev);
-	bus_node->count_cells(dev, &na, &ns);
+	na = of_simple_addr_cells(dev);
+	ns = of_simple_size_cells(dev);
```
You've added DT example - thank you. But still what is wrong - what's src address, translated dma address? is problem related to RAM or IO? which translation step gives incorrect result? 

Also, pay attention that `dma-ranges = <val>`; might not need to be placed  in all these devices as it's "bus" property, in general (axi bus in this case). Have you tried to just replace `dma-ranges = <val>;` vs `dma-ranges;` in pcie/rpi1 nodes?

oleksiimoisieiev replied:
> You've added DT example - thank you. But still what is wrong - what's src address, translated dma address? is problem related to RAM or IO? which translation step gives incorrect result?
> 

As was stated in the commit description - we have to retrieve correct address-cells of the current device and then correct address-cells for the bus device.
In our case current address-cells is 2 when bus address-cells is 3. So the translation won't be messed up. This fix is about the correctness of parsing the DT parameters and not about translation itself.


> Also, pay attention that `dma-ranges = <val>`; might not need to be placed in all these devices as it's "bus" property, in general (axi bus in this case). Have you tried to just replace `dma-ranges = <val>;` vs `dma-ranges;` in pcie/rpi1 nodes?

This fix not about dma-ranges. It's about #address-cells and #size-cells. 
Testing different dma-ranges format is not in the scope of the current fix.

GrygiriiS commented on board/raspberrypi/rpi/rpi.c:
```diff
@@ -19,6 +19,8 @@
 #include <asm/arch/sdhci.h>
```
check commit message "board: raspberrypi: rpi: save board_type to the global_data", pls

oleksiimoisieiev replied:
+

GrygiriiS commented on drivers/pci/pcie_brcmstb.c:
```diff
@@ -628,6 +905,11 @@ static int brcm_pcie_of_to_plat(struct udevice *dev)
 
+	pcie->pcie_cfg = (const struct brcm_pcie_cfg_data *)dev_get_driver_data(dev);
+
+	if (IS_ENABLED(CONFIG_BCM2712) && pcie->pcie_cfg->type == BCM2712)
```
This looks a bit confusing - part of BCM2712 staff is under ifdef and part is not.

1) Seems, It all can be gracefully resolved by using just DT compatible and match_data only, especially taking into account that  you've added `enum brcm_pcie_type`.
2) if still want/have to use CONFIG_X - it should be not global, but KConfig option defined in drivers/pci/Kconfig which may depends on CONFIG_BCM2712

oleksiimoisieiev replied:
The idea was to leave CONFIG_BCM2712 option (as we have CONFIG_BCM2711 for rpi4, CONFIG_BCM2384 and so on) but make this option to indicate generic BCM2712 support. And it's purpose is to prevent unused parts to be linked to the code when BCM2712 is disabled( such ad resets in pcie or gpio in macb).
That's why this ifdef was used only in a few places. 
As for point 2 - Could you please elaborate a bit. I'm not sure I understand the purpose.

GrygiriiS replied:
The point is to have PCI_BRCMSTB_BCM2712 which belongs to "pci" and not to use some global config - tomorrow there could be BCM2713, but with the same PCIe.

And please or use ifdef everywhere or pcie->pcie_cfg->type == BCM2712 - you'r linker optimizations are partial which makes code looks very misleading. For example,` bcm2712_cfg` not under ifdef.

GrygiriiS commented on drivers/pci/pcie_brcmstb.c:
```diff
@@ -427,6 +546,111 @@ static void brcm_pcie_set_outbound_win(struct brcm_pcie *pcie,
 }
 
+#ifdef CONFIG_BCM2712
+static void brcm_pcie_resets_dt(struct udevice *dev)
```
brcm_pcie_get_resets_dt?

oleksiimoisieiev replied:
+

GrygiriiS commented on drivers/pci/pcie_brcmstb.c:
```diff
@@ -186,6 +259,52 @@ static bool brcm_pcie_rc_mode(struct brcm_pcie *pcie)
+			     RGR1_SW_INIT_1_INIT_MASK);
+}
+
+static void brcm_pcie_bridge_sw_init_set_2712(struct brcm_pcie *pcie, u32 val)
```
There you seems have 2 different reset functions for 2712 -  brcm_pcie_bridge_sw_init_set_2712() and  brcm_pcie_do_reset().

oleksiimoisieiev replied:
yes. they are. I think calling sw_init from here will not improve code readability. What do you think?

GrygiriiS commented on drivers/gpio/rp1_gpio.c:
```diff
@@ -0,0 +1,375 @@
+	u8 bank;
+	u8 offset;
+	u8 fsel;
+	u8 irq_type;
```
drop it

oleksiimoisieiev replied:
+

GrygiriiS commented on drivers/reset/reset-brcmstb.c:
```diff
@@ -0,0 +1,97 @@
+	 * operation is typically 14us for the worst case, build some slack
+	 * here.
+	 */
+	usleep_range(100, 200);
```
It's probably not sleep?

oleksiimoisieiev replied:
Sorry. What do you mean?

GrygiriiS replied:
u-boot can't sleep

oleksiimoisieiev replied:
Ah. Now i understand. That's a wrapper around delay (you can see it as define on the beginning of file. That's a common practice (as I saw similar approach on other drivers) to make a wrappers around linux kernel calls as a glue between ported code and u-boot toolchain. Just to reduce an amount of find/replace. At first I've changed all calls, but then got tired and used this way.

GrygiriiS commented on drivers/clk/Kconfig:
```diff
@@ -163,6 +163,13 @@ config CLK_OCTEON
 	help
 	  Enable this to support the clocks on Octeon MIPS platforms.
 
+config CLK_RP1
```
Probably the "drivers: clk: introduce clock driver for RP1"  should be marked as HACK, and commit msg should be definitely updated to say - it supports only NET clock.

lorc replied:
It is normal for U-Boot drivers to add only minimally needed set of features. Additional clocks can be added when they are required, driver design allows this with ease.

I can show example of exactly the same behavior in, say, QCOM drivers.

oleksiimoisieiev replied:
This is a common practice in u-boot to add only clock which are in use right now. So I think this is definitely not a HACK. As for the NET clock - the clock name is RP1_CLK_ETH_TSU. I can add it to the commit description but wasn't able to find similar mentions in the u-boot history with the list of supported clocks. Also, please note that I have added all clock devices in rp1.h.

GrygiriiS commented on drivers/net/macb.c:
```diff
@@ -142,6 +146,12 @@ struct macb_device {
 	phy_interface_t		phy_interface;
+
+#ifdef CONFIG_GPIO
+	struct gpio_desc phy_reset_gpio;
```
I have a very strong felling that NET will work without PHY gpio reset - with default settings, so neither this part, neither GPIO driver is not needed to be part of this PR.

lorc replied:
That depends on default reset state. I have a strong feeling that VPU boot code leaves PHY in reset, but @oleksiimoisieiev can correct me.

oleksiimoisieiev replied:
Unortunately - gpio driver was ported specificaly for this purpose.

GrygiriiS replied:
ok. that's sad - usually PHY are out of reset by deafult.

GrygiriiS commented on drivers/net/macb.c:
```diff
@@ -273,47 +283,69 @@ int macb_miiphy_write(struct mii_dev *bus, int phy_adr, int devad, int reg,
 
+static int macb_miiphy_reset(struct mii_dev *bus)
+{
+#if CONFIG_GPIO
```
Any way it' not u-boot style - it can be resolved basing on "compatible" and then just checked for GPIO to be valid.

oleksiimoisieiev replied:
My goal is to avoid any changes to DT. This can be implemented in different ways.

GrygiriiS replied:
It's still OPTIONAL - and  CONFIG_GPIO will not help here as it can be enabled, but phy gpio NOT provided.
There is examples in "net" how it can be solved - for example mvneta_probe()->phy_reset_gpio or  __dw_mdio_reset()

GrygiriiS commented on drivers/net/macb.c:
```diff
@@ -920,6 +968,14 @@ static int _macb_init(struct udevice *dev, const char *name)
+	 */
+	if ((IS_ENABLED(CONFIG_BCM2712)) &&
+			(gd_board_type() == BCM2712_RPI_5_B_NEW))
+		gem_init_axi(macb);
```
Pls, solve it using mach_data.

oleksiimoisieiev replied:
search for "mach_data" in entire u-boot project gives only one entry in structure vsc9953_ana_ana_tables
Maybe there is some typo in the name?

GrygiriiS replied:
```
static const struct udevice_id macb_eth_ids[] = {
	{ .compatible = "cdns,macb" },
```

But ok - they have not added bcm2712 compatible for macb, so no issue.

GrygiriiS commented on drivers/net/macb.c:
```diff
@@ -1325,6 +1408,18 @@ static int macb_eth_of_to_plat(struct udevice *dev)
 
+	if (IS_ENABLED(CONFIG_GPIO)) {
+		/* optional PHY reset-related properties */
+		ret = gpio_request_by_name(dev, "phy-reset-gpios", 0, &macb->phy_reset_gpio,
```
if it's optional you have to check for -ENOENT and dance from here on.

oleksiimoisieiev replied:
+

GrygiriiS replied:
That's not enough - see my comment above for  macb_miiphy_reset()

GrygiriiS commented on board/raspberrypi/rpi/rpi.c:
```diff
@@ -594,3 +594,26 @@ int ft_board_setup(void *blob, struct bd_info *bd)
+ * RP1 pci device should be initialized by the PCI subsystem because
+ * it is under develop right now and depends from the final device-tree
+ * format from the Linux Kernel. Current device-tree format violates
+ * pci driver model. So this should be changed after upstreaming RP1
```
It's misleading - this commit is needed for u-boot - but commit message and this comment are talking about "Linux Kernel"?

oleksiimoisieiev replied:
because format was derived from the Linux Kernel. And I'm stating that we are dependent from the Linux kernel driver development state.

GrygiriiS commented on drivers/mfd/rp1.c:
```diff
@@ -63,6 +63,20 @@ static int rp1_probe(struct udevice *dev)
+	 */
+	dm_pci_write_config32(dev, PCI_BASE_ADDRESS_1, 0);
+	dm_pci_write_config32(dev, PCI_BASE_ADDRESS_0, 0x40000);
+
```
So my understanding is that it is required for Linux to probe PCIe and RP1, right? Wouldn't reset in Linux or here in u-boot help with that?

oleksiimoisieiev replied:
Because Linux expecting PCIE to be configured in some state after firmware initialization. And doesn't reconfigure bars. It was lucky coincidence that Linux does initialization the same way firmware sets the values. but u-boot does init in the different pattern. that's why we see this HACK. Why it is so - well I can't tell you. I have no documentation for PCI 2712

lorc replied:
Yeah, design of RP1 driver on Linux is ... subpar.  It is working because of sheer luck (or because of assumption that prior bootloader will configure device in a certain way).

lorc commented on configs/rpi_arm64_defconfig:
```diff
@@ -56,3 +57,28 @@ CONFIG_SYS_WHITE_ON_BLACK=y
+
+CONFIG_RP1_GPIO=y
+CONFIG_BOARD_LATE_INIT=y
+CONFIG_BOARD_TYPES=y
```
Missing EOL

oleksiimoisieiev replied:
+

lorc commented on include/broadcom/bcm_board_types.h:
```diff
@@ -0,0 +1,18 @@
+#ifndef _BCM_BOARD_TYPES_H_
```
Should you add copyright information here?

oleksiimoisieiev replied:
Yeah. right...

lorc commented on drivers/clk/clk-rp1.c:
```diff
@@ -0,0 +1,495 @@
+
+#include <dt-bindings/clk/rp1.h>
+
+#define PLL_SYS_CS			0x08000
```
You can safely remove most of those defines as they are not used by the driver.

oleksiimoisieiev replied:
removed

lorc reviewed (commented):
@GrygiriiS did a great review, I have a little to add from my side

lorc commented on configs/rpi_arm64_defconfig:
```diff
@@ -56,3 +57,21 @@ CONFIG_SYS_WHITE_ON_BLACK=y
 CONFIG_PHYS_TO_BUS=y
+
+CONFIG_RESET_BRCMSTB=y
+CONFIG_MFD_RP1=y
```
In the very first commit you are referencing features that are not available yet. Please move this patch to the very end of series. Or add new defconfig entries gradually, when implementing corresponding features

oleksiimoisieiev replied:
fixed

lorc commented on arch/arm/mach-bcm283x/init.c:
```diff
@@ -76,15 +76,26 @@ static struct mm_region bcm2712_mem_map[MEM_MAP_MAX_ENTRIES] = {
 		.attrs = PTE_BLOCK_MEMTYPE(MT_NORMAL) |
 			 PTE_BLOCK_INNER_SHARE
-	}, {
+	},
```
Why are you changing code style here?

oleksiimoisieiev replied:
That wasn't me ))). Fixed

lorc commented on arch/arm/mach-bcm283x/init.c:
```diff
@@ -76,15 +76,26 @@ static struct mm_region bcm2712_mem_map[MEM_MAP_MAX_ENTRIES] = {
 			 PTE_BLOCK_NON_SHARE |
 			 PTE_BLOCK_PXN | PTE_BLOCK_UXN
-	}, {
+	},
```
... and here

oleksiimoisieiev replied:
still wasn't me ))) fixed.

lorc commented on drivers/pci/pcie_brcmstb.c:
```diff
@@ -186,6 +246,52 @@ static bool brcm_pcie_rc_mode(struct brcm_pcie *pcie)
+	void __iomem *base = pcie->base;
+	u32 val = readl(base + PCIE_MISC_PCIE_STATUS);
+
+	return !!FIELD_GET(STATUS_PCIE_PORT_MASK_2712, val) | 1;
```
Why ` | 1`? Is this some leftover from your debug code?

Also no need to use `!!` here.

oleksiimoisieiev replied:
that's because FIELD_GET returns 0... Looks like a dirty hack to make pci work even if pcie_status is 0.
Or probably (and most likely) STATUS_PCIE_PORT_MASK_2712 is wrong. But I can only guess because I have no documentation.

lorc replied:
This worth investigation and some comment at least.

oleksiimoisieiev replied:
Switched to brcm_pcie_rc_mode and remove special implementation for 2712. looks like it is a mistake because it is not likely that there it unlikely that port mask was changed in the same controller. That's why on the linux code this hack was applied.

lorc commented on drivers/pci/pcie_brcmstb.c:
```diff
@@ -427,6 +533,109 @@ static void brcm_pcie_set_outbound_win(struct brcm_pcie *pcie,
+				   0x1600);
+	for (i = 0; i < ARRAY_SIZE(regs); i++) {
+		brcm_pcie_mdio_read(pcie->base, MDIO_PORT0, regs[i], &tmp);
+		printf("PCIE MDIO pre_refclk 0x%02x = 0x%04x\n",
```
Should it be hidden under some `DEBUG` macro?

oleksiimoisieiev replied:
fixed

lorc commented on drivers/pci/pcie_brcmstb.c:
```diff
@@ -427,6 +533,109 @@ static void brcm_pcie_set_outbound_win(struct brcm_pcie *pcie,
+	for (i = 0; i < ARRAY_SIZE(regs); i++) {
+		brcm_pcie_mdio_write(pcie->base, MDIO_PORT0, regs[i], data[i]);
+		brcm_pcie_mdio_read(pcie->base, MDIO_PORT0, regs[i], &tmp);
+		printf("PCIE MDIO post_refclk 0x%02x = 0x%04x\n",
```
Should it be hidden under some `DEBUG` macro?

oleksiimoisieiev replied:
fixed

lorc commented on drivers/pci/pcie_brcmstb.c:
```diff
@@ -427,6 +533,109 @@ static void brcm_pcie_set_outbound_win(struct brcm_pcie *pcie,
+		return;
+
+	/* program additional inbound windows (RC_BAR4..RC_BAR10) */
+	count = 7;
```
So, you define `count` only to be used as a constant in `for` loop below?

oleksiimoisieiev replied:
+

lorc commented on drivers/pci/pcie_brcmstb.c:
```diff
@@ -442,34 +651,56 @@ static int brcm_pcie_probe(struct udevice *dev)
-	setbits_le32(base + PCIE_RGR1_SW_INIT_1,
-		     RGR1_SW_INIT_1_INIT_MASK | RGR1_SW_INIT_1_PERST_MASK);
+	if (pcie->pcie_cfg->type != BCM2712) {
+		pcie->pcie_cfg->bridge_sw_init_set(pcie, 1);
```
If you already have callbacks defined, why check for `type`? Just define an empty callbacks for `BCM2712`

oleksiimoisieiev replied:
It's not empty for BCM2712, it's brcm_pcie_bridge_sw_init_set_2712
I just don't want to call this callback for BCM2712. Again - I can't tell why ((

lorc replied:
Well, then it worth investigation or `TODO` at least.

oleksiimoisieiev replied:
I did refactoring and updated calls

lorc commented on drivers/pci/pcie_brcmstb.c:
```diff
@@ -504,12 +764,14 @@ static int brcm_pcie_probe(struct udevice *dev)
 	writel(0xffffffff, base + PCIE_MSI_INTR2_CLR);
 
+	if (pcie->pcie_cfg->type == BCM2712)
+		brcm_pcie_window_prog(dev);
```
`brcm_pcie_window_prog` already checks for `type`

oleksiimoisieiev replied:
thanks

lorc commented on drivers/pci/pcie_brcmstb.c:
```diff
@@ -442,34 +651,56 @@ static int brcm_pcie_probe(struct udevice *dev)
-		     RGR1_SW_INIT_1_INIT_MASK | RGR1_SW_INIT_1_PERST_MASK);
+	if (pcie->pcie_cfg->type != BCM2712) {
+		pcie->pcie_cfg->bridge_sw_init_set(pcie, 1);
+		pcie->pcie_cfg->perst_set(pcie, 1);
```
You are calling this callback below, but you are not checking for `type` here

oleksiimoisieiev replied:
Let me describe how and why this was done:
these 2 pieces of code before and after sleep is bridge reset. for BCM2712 both bridge and rescal resets should be done. That's why it looks like this:
for the boards != BCM2712 we call bridge_sw_init_set and perst set, then after sleep bridge_sw_init_set.
for BCM2712 we call brcm_pcie_do_reset. which does basically bridge_sw_init, but with rescal reset deassertion.
From my standpoint I did the best I could to minimize impact of the already written code and make more or less understandable implementation for bcm2712

lorc replied:
> From my standpoint I did the best I could to minimize impact of the already written code and make more or less understandable implementation for bcm2712

That the problem. Do not try to minimize impact to existing code. In the end the driver will consist of `ifdef`/`if` spaghetty which is really hard try to understand.  Try to make final code clear. Even if it means that you need to rewrite half of the original driver. This is okay.

oleksiimoisieiev replied:
updated

lorc commented on drivers/clk/clk-rp1.c:
```diff
@@ -0,0 +1,280 @@
+		 data->name, rate, 0x2faf080))
+		div = 1 << CLK_DIV_FRAC_BITS;
+
+	spin_lock(&clockman->regs_lock);
```
I believe you can remove all spinlocks, as U-Boot is purely single-threaded

oleksiimoisieiev replied:
I know. But this is already handled in compat.h, which purpose is to allow developers to port linux code with minimal changes.
spin_lock define is the following:
#define spin_lock(lock) do {} while (0)

lorc reviewed (commented):
@GrygiriiS, may I ask you to resolve your comments if they are addressed by @oleksiimoisieiev ? 

@oleksiimoisieiev could you please respond to @GrygiriiS's comments?

I am asking because I don't quite understand what is left to be done.

oleksiimoisieiev commented:
> @GrygiriiS, may I ask you to resolve your comments if they are addressed by @oleksiimoisieiev ?
> 
> @oleksiimoisieiev could you please respond to @GrygiriiS's comments?
> 
> I am asking because I don't quite understand what is left to be done.

We went through all comments from @GrygiriiS on wed.

lorc reviewed (approved):
Okay, no more comments from my side

`Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>`

oleksiimoisieiev commented:
Info from @GrygiriiS  - he is ok with changes. this can be merged. @lorc
